org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/pom/ReImportMavenProjectsHandler.java

```diff
@@ -0,0 +1,179 @@
+import org.eclipse.lsp4j.jsonrpc.json.adapters.CollectionTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EnumTypeAdapterFactory;
+
+/** 
+ * Command to update maven projects.
+ * 
+ * @author Mykola Morhun
+ */
+public class ReImportMavenProjectsHandler {
+  public static long REIMPORT_TIMEOUT = 60_000L;
+
+  private static final Gson gson =
+      new GsonBuilder()
+          .registerTypeAdapterFactory(new CollectionTypeAdapterFactory())
+          .registerTypeAdapterFactory(new EitherTypeAdapterFactory())
+          .registerTypeAdapterFactory(new EnumTypeAdapterFactory())
+          .create();
+
+  private static final PreferenceManager preferenceManager = new PreferenceManager();
+  private static final ProjectsManager projectsManager = new ProjectsManager(preferenceManager);
+  private static final IJobManager jobManager = Job.getJobManager();
+  private static final JobChangedListener jobChangedListener = new JobChangedListener();
+  private static final Lock lock = new ReentrantLock();
+
+  private static CountDownLatch jobsToBeFinished;
+  // name -> path
+  private static Map<String, String> projectsToBeUpdated = new ConcurrentHashMap<>();
+  // set of paths to updates projects; path -> ""
+  private static Map<String, String> updatedProjectsPaths = new ConcurrentHashMap<>();
+
+  /**
+   * Updates given maven projects.
+   *
+   * @param arguments contains ReImportMavenProjectsCommandParameters in the first element
+   * @param progressMonitor progress monitor
+   * @return URIs of updated projects
+   */
+  public static List<String> reImportMavenProjects(
+      List<Object> arguments, IProgressMonitor progressMonitor) {
+
+    if (lock.tryLock()) {
+      try {
+        updatedProjectsPaths.clear();
+
+        ReImportMavenProjectsCommandParameters parameters =
+            gson.fromJson(
+                gson.toJson(arguments.get(0)), ReImportMavenProjectsCommandParameters.class);
+
+        ensureNotCancelled(progressMonitor);
+        return new ArrayList<>(
+            doReImportMavenProjects(parameters.getProjectsToUpdate(), progressMonitor));
+      } catch (InterruptedException e) {
+        throw new OperationCanceledException(e.getMessage());
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    throw new RuntimeException("Reimport is already in progress.");
+  }
+
+  // TODO this is temporary solution.
+  // It shouldn't block response but send back projects to which a job was submitted.
+  // Then progress should be send separately.
+  // For details see https://github.com/eclipse/eclipse.jdt.ls/issues/404
+  private static Set<String> doReImportMavenProjects(
+      List<String> projectsUri, IProgressMonitor progressMonitor) throws InterruptedException {
+    final List<IProject> projects = validateProjects(projectsUri, progressMonitor);
+    jobManager.addJobChangeListener(jobChangedListener);
```

Comment:
If something goes wrong after this point, you're never removing the jobChangeListener. Try-finally, please.

Reply:
nice tip